pkg/volume/gce_pd/gce_util.go

```diff
@@ -122,9 +137,111 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
 	if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
 		return err
 	}
+
+	devicePaths := []string{
+		path.Join("/dev/disk/by-id/", "google-"+pd.pdName),
+		path.Join("/dev/disk/by-id/", "scsi-0Google_PersistentDisk_"+pd.pdName),
+	}
+
+	if pd.partition != "" {
+		for i, path := range devicePaths {
+			devicePaths[i] = path + "-part" + pd.partition
+		}
+	}
+	numTries := 0
+	// Wait for the disk device to detach
+	for {
+		pathExists := false
+		for _, path := range devicePaths {
+			udevadmChangeToDrive(path)
+			_, err := os.Stat(path)
+			if err == nil {
+				pathExists = true
+			}
+			if err != nil && !os.IsNotExist(err) {
+				return err
+			}
+		}
+		if !pathExists {
+			break
+		}
+		numTries++
+		if numTries == 10 {
+			return errors.New("Could not detach disk: Timeout after 10s")
+		}
+		time.Sleep(time.Second)
+	}
+
 	return nil
 }
 
+// Calls "udevadm trigger --action=change" for newly created "/dev/sd*" drives (exist only in after set).
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToNewDrives(sdBefore, sdAfter []string) {
+	sdexists := make(map[string]bool)
+	for _, sd := range sdBefore {
+		sdexists[sd] = true
+	}
+	for _, sd := range sdAfter {
+		if !sdexists[sd] {
+			udevadmChangeToDrive(sd)
+		}
+	}
+}
+
+// Calls "udevadm trigger --action=change" on the specified drive.
+// drivePath must be the the block device path to trigger on, in the format "/dev/sd*", or a symlink to it.
+// This is workaround for Issue #7972. Once the underlying issue has been resolved, this may be removed.
+func udevadmChangeToDrive(drivePath string) {
+	glog.V(5).Infof("udevadmChangeToDrive: drive=%q", drivePath)
+
+	// Evaluate symlink, if any
+	drive, err := filepath.EvalSymlinks(drivePath)
+	if err != nil {
+		glog.Errorf("udevadmChangeToDrive: filepath.EvalSymlinks(%q) failed with %v.", drivePath, err)
+		return
+	}
+	glog.V(5).Infof("udevadmChangeToDrive: symlink path is %q", drive)
+
+	// Check to make sure input is "/dev/sd*"
+	if !strings.Contains(drive, "/dev/sd") {
+		glog.Errorf("udevadmChangeToDrive: expected input in the form \"/dev/sd*\" but drive is %q.", drive)
+		return
+	}
+
+	// Get the DEVPATH for the specified block device
+	udevInfoOutBytes, err := exec.New().Command("udevadm", "info", "--q=all", fmt.Sprintf("--name=%s", drive)).CombinedOutput()
```

Comment:
So I've poked around a bit and I think that we can remove most of this code.
example:
$ /sbin/udevadm trigger --action=change --verbose --property-match=DEVNAME=/dev/sd
b
/sys/devices/pci0000:00/0000:00:03.0/virtio0/host0/target0:0:2/0:0:2:0/block/sdb

So basically you can remove the "udevadmin info" invocation, and all the stuff to parse out DEVPATH. 

And the "udevadmin trigger" command reduces to just:

udevadmin trigger --action=change --property-match=DEVNAME=$drive